cmd/kubeadm/app/util/apiclient/wait.go

```diff
@@ -208,34 +209,51 @@ func (w *KubeWaiter) WaitForStaticPodControlPlaneHashes(nodeName string) (map[st
 func (w *KubeWaiter) WaitForStaticPodSingleHash(nodeName string, component string) (string, error) {
 
 	componentPodHash := ""
-	var err error
+	var err, lastErr error
 	err = wait.PollImmediate(kubeadmconstants.APICallRetryInterval, w.timeout, func() (bool, error) {
 		componentPodHash, err = getStaticPodSingleHash(w.client, nodeName, component)
 		if err != nil {
+			lastErr = err
 			return false, nil
 		}
 		return true, nil
 	})
 
+	if err != nil {
+		err = errors.Wrapf(lastErr, "failed to obtain static Pod hash for component %s on Node %s", component, nodeName)
+	}
 	return componentPodHash, err
 }
 
 // WaitForStaticPodHashChange blocks until it timeouts or notices that the Mirror Pod (for the Static Pod, respectively) has changed
 // This implicitly means this function blocks until the kubelet has restarted the Static Pod in question
 func (w *KubeWaiter) WaitForStaticPodHashChange(nodeName, component, previousHash string) error {
-	return wait.PollImmediate(kubeadmconstants.APICallRetryInterval, w.timeout, func() (bool, error) {
+	var err, lastErr error
+	err = wait.PollImmediate(kubeadmconstants.APICallRetryInterval, w.timeout, func() (bool, error) {
 
 		hash, err := getStaticPodSingleHash(w.client, nodeName, component)
 		if err != nil {
+			lastErr = err
 			return false, nil
 		}
+		// Set lastErr to nil which identities `getStaticPodSingleHash` has no err already, and we can just care other err if any
+		lastErr = nil
 		// We should continue polling until the UID changes
 		if hash == previousHash {
 			return false, nil
 		}
 
 		return true, nil
 	})
+
+	// distinguish `getStaticPodSingleHash` err and kubelet possible err here
+	if lastErr != nil {
+		return errors.Wrapf(lastErr, "failed to obtain static Pod hash for component %s on Node %s", component, nodeName)
```

Comment:
so we seem to be repeating this text in a number of places.

instead we should modify the function and just return lastErr in all the places.

```
// getStaticPodSingleHash computes hashes for a single Static Pod resource
func getStaticPodSingleHash(client clientset.Interface, nodeName string, component string) (string, error) {
	staticPodName := fmt.Sprintf("%s-%s", component, nodeName)
	staticPod, err := client.CoreV1().Pods(metav1.NamespaceSystem).Get(context.TODO(), staticPodName, metav1.GetOptions{})
	if err != nil {
		return "", errors.Wrapf(lastErr, "failed to obtain static Pod hash for component %s on Node %s", component, nodeName)
	}
	staticPodHash := staticPod.Annotations["kubernetes.io/config.hash"]
	return staticPodHash, nil
}
```

i've also removed the Printf in the function

Reply:
done here, too.